File: backend/diagnose_mongo.py
import dns.resolver
import socket
import ssl
import sys

def check_dns(hostname):
    print(f"--- Checking DNS for {hostname} ---")
    try:
        # Check SRV record for Atlas
        srv_name = f"_mongodb._tcp.{hostname}"
        answers = dns.resolver.resolve(srv_name, 'SRV')
        for rdata in answers:
            print(f"SRV Target: {rdata.target}")
    except Exception as e:
        print(f"DNS SRV Lookup Failed: {e}")

    try:
        # Check A record
        ip = socket.gethostbyname(hostname)
        print(f"A Record: {ip}")
    except Exception as e:
        print(f"DNS A Record Lookup Failed: {e}")

def check_port(hostname, port):
    print(f"--- Checking Port {port} on {hostname} ---")
    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(5)
        result = sock.connect_ex((hostname, port))
        if result == 0:
            print(f"Port {port} is OPEN")
        else:
            print(f"Port {port} is CLOSED (Code: {result})")
    except Exception as e:
        print(f"Port Check Error: {e}")
    finally:
        sock.close()

if __name__ == "__main__":
    atlas_host = "cluster0.dpdkhhz.mongodb.net"
    check_dns(atlas_host)
    check_port(atlas_host, 27017)
    # Also check a direct shard node if known from previous attempts
    shard0 = "ac-kscv88c-shard-00-00.dpdkhhz.mongodb.net"
    check_port(shard0, 27017)

File: backend/diagnose_mongo.js
import { MongoClient } from 'mongodb';
import dns from 'dns';
import net from 'net';

const hostname = 'cluster0.dpdkhhz.mongodb.net';
const srvName = `_mongodb._tcp.${hostname}`;

console.log(`--- Starting Diagnostics for ${hostname} ---`);

// 1. Check DNS resolution for SRV
dns.resolveSrv(srvName, (err, addresses) => {
    if (err) {
        console.error(`❌ DNS SRV Resolve Failed: ${err.message}`);
    } else {
        console.log(`✅ DNS SRV Resolved:`, addresses);
        
        // 2. Check Port for resolved addresses
        addresses.forEach(addr => {
            const client = new net.Socket();
            client.setTimeout(5000);
            console.log(`Checking port 27017 on ${addr.name}...`);
            client.connect(27017, addr.name, () => {
                console.log(`✅ Successfully connected to ${addr.name}:27017`);
                client.destroy();
            });
            client.on('error', (err) => {
                console.error(`❌ Port connection failed on ${addr.name}: ${err.message}`);
            });
            client.on('timeout', () => {
                console.error(`❌ Port connection timeout on ${addr.name}`);
                client.destroy();
            });
        });
    }
});

// 3. Test MongoDB client connection separately with a timeout
const uri = 'mongodb+srv://ns300142_db_user:ns123%40123@cluster0.dpdkhhz.mongodb.net/advantage-gen?retryWrites=true&w=majority&appName=Cluster0';
const client = new MongoClient(uri, { serverSelectionTimeoutMS: 5000 });

async function testConnection() {
    try {
        console.log(`--- Testing Connection with MongoClient (Timeout 5s) ---`);
        await client.connect();
        console.log("✅ SUCCESS: MongoClient connected to Atlas!");
        await client.db("admin").command({ ping: 1 });
        console.log("✅ SUCCESS: Database pinged successfully.");
    } catch (err) {
        console.error(`❌ MongoClient Connection Failed: ${err.message}`);
    } finally {
        await client.close();
    }
}

testConnection();
